torch/csrc/jit/codegen/cuda/lower_alias_memory.cpp

```diff
@@ -61,127 +63,903 @@ class SymbolicSizePrinter : private kir::IrVisitor {
+      handle(for_loop);
+    } else if (auto ite = dynamic_cast<const kir::IfThenElse*>(node)) {
+      handle(ite);
+    } else {
+      indent();
+      ir_printer_.printNode(node);
+    }
+    if (auto alloc = dynamic_cast<const kir::Allocate*>(node)) {
+      printAllocInfo(alloc);
+    }
+  }
+
+  void handle(const kir::ForLoop* node) {
+    indent();
+    os_ << "FOR ";
+    ir_printer_.printNode(node->index());
+    os_ << " in ";
+    ir_printer_.printNode(node->iter_domain());
+    os_ << ":\n";
+  }
+
+  void handle(const kir::IfThenElse* node) {
+    // This pass doesn't yet need to handle
+    //  ite but could fill in the blank here
+    //  if this printer can be used for
+    //  other passes or we have more
+    //  complex ite pattern.
+    TORCH_INTERNAL_ASSERT(false, "unsupported");
+  }
+
+  void printAllocInfo(const kir::Allocate* alloc);
+
+  std::stringstream& indent() {
+    for (int i = 0; i < indent_level_; i++) {
+      os_ << "  ";
+    }
+    return os_;
+  }
+
+  void startBlock() {
+    indent_level_++;
+  }
+
+  void endBlock() {
+    indent_level_--;
+  }
+
+ private:
+  std::stringstream os_;
+  kir::IrPrinter ir_printer_;
+  int indent_level_ = 0;
+
+  std::vector<DebugEntryPtr> debug_info_;
+  BufferUseDefInfo* buffer_info_ = nullptr;
+};
+
+//! Utility class for modeling the liveness interval.
+//! The first write and last read
+//! is based on the position on the linear order within
+//! the Kernel IR.
+//!  The interval is semi-open,
+//!     i.e. [First_Write, Last_Read)
+//!  So the buffer is NOT available at exactly First_Write
+//!   position while it IS available at Last_Read.
+class BufferLiveInterval {
+ public:
+  // Simple detection of intersection of two intervals
+  bool intersect(BufferLiveInterval* other) {
+    if (first_write_pos_ <= other->first_write_pos_) {
+      return other->first_write_pos_ < last_read_pos_;
+    } else {
+      return first_write_pos_ < other->last_read_pos_;
+    }
+  }
+
+  void markWrite(int pos) {
+    if (first_write_pos_ == -1) {
+      first_write_pos_ = pos;
+    }
+  }
+
+  void markRead(int pos) {
+    last_read_pos_ = pos;
+    TORCH_INTERNAL_ASSERT(
+        pos > first_write_pos_, "marking a read before write");
+    all_read_pos_.push_back(pos);
+  }
+
+  const auto& allReads() {
+    return all_read_pos_;
+  }
+
+  auto firstWrite() const {
+    return first_write_pos_;
+  }
+
+  auto lastRead() const {
+    return last_read_pos_;
+  }
+
+  std::string toString() {
+    std::stringstream ss;
+    ss << "[ " << first_write_pos_ << " , " << last_read_pos_ << " )";
+    return ss.str();
+  }
+
+ private:
+  int first_write_pos_ = -1;
+  int last_read_pos_ = -1;
+  std::vector<int> all_read_pos_;
+};
+
+using BufferLiveIntervalPtrList = std::vector<BufferLiveInterval*>;
+
+//! Thin struct to keep track of loops. The actual loop body is
+//!  considered live in [start_pos, end_pos)
+struct ScopeInfo {
```

Comment:
The expr counter is absolute on the flattened order. `resetExprCounter` is called once in `collectScopeUseDefInfo` and once in `collectScopeInfo` because they're two separate passes through the whole structure and expr counter was used to keep track of the line number of the expression being handled. This method is valid in this pass because this pass doesn't really insert/delete any expression.

Maybe in a follow up it is worthwhile to have a `line_number` attribute introduced in kernel ir so this part could be simplified.

Reply:
Oof okay, must have misread this. Line number could be nice for debugging compilation issues. nvrtc print out is very long with all our inserted code, we could try to chop that out and number the lines starting from the end when we print the error. However, seems out of scope from this pr.

Reply:
Sure we could try to label line numbers in codegen phase as well. It sounds helpful.